### ISSUE TITLE
Certificate validation callback per ConnectionSettings instance

### ISSUE THREAD

**Issue by george-polevoy** [elastic/elasticsearch-net#2620]
**Describe the feature**:

Currently, I have to write following to ignore a self-signed certificate for elc on localhost:

ServicePointManager.ServerCertificateValidationCallback = (sender,certificate,chain,sslPolicyErrors)=>{
    var uriString = ... something like ((HttpWebRequest)sender).Uri.ToString()
    if (uriString.StartsWith("https://localhost:9200"))
         return true;
    else if (... // there are many different REST api's and I don't want to ignore the certificates for all of them.

The feature I'm requesting would allow following configuration:

new ConnectionSettings(connectionPool)
 // first three lines are existing features of ConnectionSettings fluent API:
                .BasicAuthentication("elastic", "*******")
                .RequestTimeout(TimeSpan.FromSeconds(5))
                .PrettyJson()
// the following line is the new feature I'm requesting:
                .ServerCertificateCallback((sender,certificate,chain,sslPolicyErrors)=>return true);

TL;DR

For a client accessing different resources using HTTP protocol and SSL/TLS, it's important to control server certificate validation behaviour. A typical example would be a accessing an HTTP resource with a self-signed certificate.

It's possible to specify Server Certificate Validation Callback per AppDomain using ServicePointManager.ServerCertificateValidationCallback property:
https://msdn.microsoft.com/en-us/library/system.net.servicepointmanager.servercertificatevalidationcallback(v=vs.110).aspx

One could just use:

ServicePointManager.ServerCertificateValidationCallback = (sender,certificate,chain,sslPolicyErrors)=>return true;

to allow a self-signed certificate, but that would ignore certificate errors for all other resources too in this process/appdomain

This is fine for simple scenarios, but sometimes different settings should be specified for different resources. Some processes could integrate with dozens of resources, and it's awkward to specify a 'one size fits all' callback per entire app domain or process, as that would require a complex conditional logic inside this callback delegate.

It is much easier to override server certificate validation behavior on per-feature basis.
For example, one would like to ignore all certificate errors for local elastic search instance with self-signed certificate, but still validate some other resource's certificate.
In this case it's more convenient to set the callback per request:

https://msdn.microsoft.com/en-us/library/system.net.httpwebrequest.servercertificatevalidationcallback(v=vs.110).aspx

**Comment by russcam** [comment]
Thanks for submitting the issue @george-polevoy 👍 Would you be interested in submitting a PR for this?

**Comment by george-polevoy** [comment]
@russcam 
While working on the PR, I discovered a possibility to do it in existing API.

There is a way of getting to HttpWebRequest through the `HttpConnection.CreateHttpWebRequest` virtual method, and there is a constructor accepting an IConnection, so one can derive a class from `HttpConnection` and 
   
```C#
private class MyCertificateIgnoringHttpConnection : HttpConnection
{
    protected override HttpWebRequest CreateHttpWebRequest(RequestData requestData)
    {
        var httpWebRequest = base.CreateHttpWebRequest(requestData);
        httpWebRequest.ServerCertificateValidationCallback = (sender, certificate, chain, errors) => true;
        return httpWebRequest;
    }
}
```

Now you can create your ConnectionSettings, providing this new class as an IConnection parameter to the constructor, such as in the following example:

```C#
var node = new Uri("https://localhost:9200");
var connectionPool = new SniffingConnectionPool(new[] {node});
var connectionSettings = new ConnectionSettings(connectionPool, new MyCertificateIgnoringHttpConnection());
```

**Comment by Mpdreamz** [comment]
@george-polevoy good find :) This has always been our suggested way to do this, and a big reason why we made those methods virtual.

The original reason for not exposing on `ConnectionSettings` is that we do not want to abstract too much over the given options that are available in both WebRequest (which the desktop CLR HttpConnection uses) and HttpClient (which the core CLR client uses).

Revisiting this PR myself now though and adding `ServerCertificateValidationCallback` on `ConnectionSettings` so that it works both for desktop CLR and core CLR as its a super common use case. 

Will also add helpers that allow you to easily pass your custom cert and create a callback handler for it as can be very complex see e.g:

https://www.elastic.co/guide/en/elasticsearch/client/net-api/1.x/security.html#trusting-certifcate-example

**Comment by russcam** [comment]
:+1: I was going to suggest exposing this on ConnectionSettings :smile:

**Comment by Mpdreamz** [comment]
This is now supported on `ConnectionSettings` see our documentation for it here:

https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/working-with-certificates.html#_client_certificates
